README.md
# Replacing-John-Doe
## What does it do?
It does what the names says it will do: it replaces John Doe's. Replacing John Doe (RJD) is an API that generates more believable placeholder data so that you do not have to go back to Lorem Ipsum's and John Doe's.  
## Why is it useful?
When you are developing, it can be very helpful to have, for example, a database with data that you can work on. You probably don't want to work on the real data when you're just developing - or you might not have real data yet. This API strives to give you an easy way to quickly have a lot of data to test with, without adding extra work. 
## Getting Started
### Running Locally
You can run the api locally using Docker. If you do not have Docker installed, please refer to [the Docker documentation](https://docs.docker.com/). 
1. Create a container using the docker-compose file; run `docker-compose up` in your terminal.
2. Navigate to the api in your browser using the url: `localhost:3000`. 

### Endpoints
1. **/createPlaceholderData** POST: /create/:type/:data -> type is either 0 (title), 1 (name) or 2 (text); data is a string. Saves data to the database. Returns the uuid of the newly created data.
2. **/getPlaceholderData** GET: /getData/:type -> type is either 0 (title), 1 (name) or 2 (text). Returns a random piece of data from the given type.
3. **/deletePlaceholderData** GET: /delete/:uuid -> deletes the data with the given uuid
4. **/updatePlaceholderData** GET: /update/:uuid/:type/:data -> edits the data with given uuid to have the given type and data
5. **/getAllPlaceholderData** GET: /getAllPlaceholderData -> returns a json with all of the placeholder data currently in the database
### Tests
1. **Endpoints**: testing the various endpoints.
2. **Helpers**: tests the helper functions (e.g. checkDataLength function)
3. **Integration**: runs through the endpoints to see how they work together

**Running the tests**
1. Navigate to the api folder
2. enter `npm test` in the terminal
## Where can I find help?
If you have a question, please email me at *sarah.verbelen@student.ehb.be*.
## Project Status
This project is currently **in development**.
## Authors
* Sarah Verbelen

## Github Flow vs GIT Flow
This project uses Github flow instead of GIT Flow. The reason for this is that I want to keep this project simple and sweet and the rigid branch structure of a traditional GIT Flow would needlessly complicate things. Additionally, as I am only a single developer working on this project, there is no need for the overview that the GIT Flow provides.

CHANGELOG.md
# Changelog
All notable changes will be documented here.

## [1.0.1] - 15/01/2021
### Added
* GetAllPlaceholderData Endpoint
### Changed
* integration test now tests if deleted data still exists
### Fixed

## [1.0.0] - 05/01/2021
First release.

api/src/__tests__/helpers.test.js
let Helpers = require('../utils/helpers.js');

describe('testing the checkDataLength function', () => {
    test('data has to be a string', () => {
        expect(Helpers.checkDataLength(5)).toBe(false);
    });

    test('data has to be longer than one', () => {
        expect(Helpers.checkDataLength("h")).toBe(false);
    });

    test('data has to be shorter than 25', () => {
        expect(Helpers.checkDataLength("012345678901234567890123456789")).toBe(false);
    });

    test('data between 2 and 24 length that is a string is good', () => {
        expect(Helpers.checkDataLength('hallo')).toBe(true);
    })
});

api/src/__tests__/endpoints.test.js
const supertest = require('supertest');
const app = require('./../server.js');

const request = supertest(app);


describe('GET the test endpoint', () => {
    test('/test should respond with statuscode 200', async (next) => {
        try {
            setTimeout(async function () {
                const response = await request.get('/test');
                expect(response.status).toBe(200);
                next();
            }, 1000);
        } catch (e) {}
    });
});


describe('GET the test endpoint', () => {
    test('/test should respond with statuscode 200', async (next) => {
        try {
            const response = await request.get('/test');
            expect(response.status).toBe(200);
            next();
        } catch (e) {}
    });
});

describe('create placeholder data', () => {
    test('POST request /createPlaceholderData cannot be empty', async (next) => {
        try {
            const response = await request.post('/createPlaceholderData');
            expect(response.status).toBe(400);
            next();
        } catch (e) {
            // console.log(e);
        }
    });

    test('POST request /createPlaceholderData cannot send only data', async (next) => {
        try {
            const response = await request.post('/createPlaceholderData/test1234');
            expect(response.status).toBe(400);
            next();
        } catch (e) {}
    });

    test('POST request /createPlaceholderData cannot send only a type', async (next) => {
        try {
            const response = await request.post('/createPlaceholderData/2');
            expect(response.status).toBe(400);
            next();
        } catch (e) {}
    });

    test('POST request /createPlaceholderData has to have data and a type', async (next) => {
        try {
            const response = await request.post('/createPlaceholderData/2/test1234');
            expect(response.status).toBe(200);
            next();
        } catch (e) {}
    });


    test('POST request /createPlaceholderData: the parameter -type- has to be either 0, 1 or 2', async (next) => {
        try {
            const response = await request.post('/createPlaceholderData/2/hello');
            expect(response.status).toBe(200);
            next();
        } catch (e) {}
    });

    test('POST request /createPlaceholderData: the parameter -type- cannot be something that isnt 0, 1 or 2', async (next) => {
        try {
            const response = await request.post('/createPlaceholderData/5/hello');
            expect(response.status).toBe(400);
            next();
        } catch (e) {}
    });
});

describe('GET: /getPlaceholderData endpoint', () => {
    test('/getData cannot be called without parameters', async (next) => {
        try {
            const response = await request.get('/getPlaceholderData');
            expect(response.status).toBe(400);
            next();
        } catch (e) {}
    });

    test('/getPlaceholderData cannot be called with a type parameter that doesnt exist (=/= 0, 1 or 2)', async (next) => {
        try {
            const response = await request.get('/getPlaceholderData/5');
            expect(response.status).toBe(400);
            next();
        } catch (e) {}
    });

    test('/getPlaceholderData called with a type parameter returns a random piece of data from that type', async (next) => {
        try {
            const response = await request.get('/getPlaceholderData/0');
            expect(response.body).toBeDefined();
            next();
        } catch (e) {}
    });
});

describe('DELETE endpoint', () => {
    test('/deletePlaceholderData cannot be called without uuid', async (next) => {
        try {
            const response = await request.get('/deletePlaceholderData');
            expect(response.status).toBe(400);
            next();
        } catch (e) {}
    });
});

describe('UPDATE endpoint', () => {
    test('/updatePlaceholderData cannot be called without parameters', async (next) => {
        try {
            const response = await request.get('/updatePlaceholderData');
            expect(response.status).toBe(400);
            next();
        } catch (e) {}
    });

    test('/updatePlaceholderData cannot be called with only 1 parameter', async (next) => {
        try {
            const response = await request.get('/updatePlaceholderData/0');
            expect(response.status).toBe(400);
            next();
        } catch (e) {}
    });

    test('/updatePlaceholderData cannot be called with only 2 parameters', async (next) => {
        try {
            const response = await request.get('/updatePlaceholderData/0/1');
            expect(response.status).toBe(400);
            next();
        } catch (e) {}
    });

});

describe('GET ALL endpoint', () => {
    test('/getAllPlaceholderData returns data', async (next) => {
        try {
            const response = await request.get('/getAllPlaceholderData');
            expect(response.status).toBe(200);
            expect(response.body).toBeDefined();
            next();
        } catch (e) {}
    });
});


api/src/__tests__/integration.test.js
const supertest = require('supertest');
const app = require('./../server.js');

const request = supertest(app);


describe('Running through the endpoints', () => {
    let uuid;
    test('creating new data', async (next) => {
        try {
            const response = await request.post('/createPlaceholderData/2/test1234');
            expect(response.status).toBe(200);
            uuid = response.body.uuid;
            console.log(uuid);
            next();
        } catch (e) {}
    });

    test('updating the data', async (next) => {
        try {
            const response = await request.get('/updatePlaceholderData/' + uuid + '/0/newtest');
            expect(response.status).toBe(200);
            next();
        } catch (e) {}
    });

    test('deleting the data', async (next) => {
        try {
            const response = await request.get('/deletePlaceholderData/' + uuid);
            expect(response.status).toBe(200);
            next();
        } catch (e) {}
    });

    test('see if the data still exists..', async (next) => {
        try {
            const response = await request.get('/getAllPlaceholderData');
            let dataDeleted = true;
            for(data in response.body.res) {
                if(data.uuid == uuid) {
                    dataDeleted = false;
                }
            }
            expect(dataDeleted).toBe(false);
            next();
        } catch (e) {}
    });
    
});

api/src/server.js
const express = require('express');
const http = require('http');
const bodyParser = require('body-parser');

const Helpers = require('./utils/helpers.js')

const app = express();
http.Server(app);

app.use(bodyParser.json());
app.use(
  bodyParser.urlencoded({
    extended: true
  })
);

// test endpoint
app.get('/test', (req, res) => {
  console.log("test");
  res.status(200).send();
});

/**
* @params: number type + string data
* @returns: statuscode 400 (error) or 200 (ok)
**/
app.post('/createPlaceholderData/:type/:data', (req, res) => {
  // check if the req has a type and data
  if (req.params.type != null && req.params.data != null) {
    if (req.params.type == 0 || req.params.type == 1 || req.params.type == 2) {
      // save the new data to the database
      const uuid = Helpers.generateUUID(); // first we generate a uuid for this placeholder data
      if (Helpers.checkDataLength(req.params.data)) {
        pg.table('placeholderData').insert({
          uuid,
          data: req.params.data,
          typeID: req.params.type
        });

        // send a statuscode when everything went right
        res.json({
          uuid: uuid
        })
      }
    }
  }
  // send an error in case the parameters are not correct / empty
  res.status(400).send();
});

// overloading the create endpoint: in case there are not enough parameters given
app.post('/createPlaceholderData/:type', (req, res) => {
  res.status(400).send();
});

app.post('/createPlaceholderData', (req, res) => {
  res.status(400).send();
});

/**
* @params: number type
* @returns: statuscode 400 (error) or json
**/
app.get('/getPlaceholderData/:type', (req, res) => {
  if (req.params.type == 0 || req.params.type == 1 || req.params.type == 2) {
    pg
      .select(['uuid', 'data', 'typeID'])
      .from('placeholderData')
      .where({
        typeID: req.params.type
      })
      .orderByRaw('random()')
      .limit(1).then(data => {
        res.json({
          res: data
        })
      });

  } else {
    res.status(400).send();
  }
});

// overloading getData endpoint in case there are no parameters given
app.get('/getPlaceholderData', (req, res) => {
  res.status(400).send();
});

/**
* @params: uuid
* @returns: statuscode 400 (error) or 200 (ok)
**/
app.get('/deletePlaceholderData/:uuid', async (req, res) => {
  await pg
    .select(['uuid', 'typeID', 'created_at', 'data'])
    .from('placeholderData')
    .where({
      uuid: req.params.uuid
    })
    .del();
  res.status(200).send();
});

// overloading delete endpoint in case there are no parameters given
app.get('/deletePlaceholderData', (req, res) => {
  res.status(400).send();
});

/**
* @params: uuid, number type, string data
* @returns: statuscode 400 (error) or 200 (ok)
**/
app.get('/updatePlaceholderData/:uuid/:type/:data', async (req, res) => {
  if (Helpers.checkDataLength(req.params.data)) {
    await pg
      .select(['uuid', 'typeID', 'created_at', 'data'])
      .from('placeholderData')
      .where({
        uuid: req.params.uuid
      })
      .update({
        data: req.params.data,
        typeID: req.params.type
      })
    res.status(200).send();
  } else {
    res.status(400).send();
  }
});

// overloading delete endpoint in case there are no parameters given
app.get('/updatePlaceholderData', (req, res) => {
  res.status(400).send();
});

app.get('/updatePlaceholderData/:uuid', (req, res) => {
  res.status(400).send();
});

app.get('/updatePlaceholderData/:uuid/:type', (req, res) => {
  res.status(400).send();
});

/**
* @params: uuid, number type, string data
* @returns: statuscode 400 (error) or 200 (ok)
**/
app.get('/getAllPlaceholderData', async (req, res) => {
  
    await pg
      .from('placeholderData')
      .join('type', 'placeholderData.typeID', 'type.typeID')
      .select(['placeholderData.uuid', 'type.type', 'placeholderData.created_at', 'placeholderData.data'])
      .where({
        uuid: req.params.uuid
      }).then(data => {
        res.json({
          res: data
        });
      })

});

// connection to the database
const pg = require('knex')({
  client: 'pg',
  version: '9.6',
  searchPath: ['knex', 'public'],
  connection: process.env.PG_CONNECTION_STRING ? process.env.PG_CONNECTION_STRING : 'postgres://example:example@localhost:5432/test'
});


/**
* @params: none
* @returns: none
**/
async function initialiseTables() {
  await pg.schema.hasTable('placeholderData').then(async (exists) => {
    if (!exists) {
      await pg.schema
        .createTable('placeholderData', (table) => {
          table.increments();
          table.uuid('uuid');
          table.string('data');
          table.integer('typeID');
          table.timestamps(true, true);
        })
        .then(async () => {
          console.log('created table placeholderData');
          // make a few titles, texts and names

          let listOfNames = ["John Doe", "Jane Doe", "Mark Verstappen", "Lily Verhaegen"];
          let listOfTitles = ["Lorem Ipsum Dolor Sit Amet", "In Het Hulst Van De Nacht", "Mijn Meest Dierbare Herinneringen"];
          let listOfTexts = ["I'm selfish, impatient and a little insecure. I make mistakes, I am out of control and at times hard to handle. But if you can't handle me at my worst, then you sure as hell don't deserve me at my best.", "Two things are infinite: the universe and human stupidity; and I'm not sure about the universe.", "Be who you are and say what you feel, because those who mind don't matter, and those who matter don't mind."];
          for (let i = 0; i < listOfNames.length; i++) {
            const uuid = Helpers.generateUUID();
            await pg.table('placeholderData').insert({
              uuid,
              data: listOfNames[i],
              typeID: 1
            })
          }

          for (let i = 0; i < listOfTitles.length; i++) {
            const uuid = Helpers.generateUUID();
            await pg.table('placeholderData').insert({
              uuid,
              data: listOfTitles[i],
              typeID: 0
            })
          }

          for (let i = 0; i < listOfTexts.length; i++) {
            const uuid = Helpers.generateUUID();
            await pg.table('placeholderData').insert({
              uuid,
              data: listOfTexts[i],
              typeID: 2
            })
          }
        });

    }
  });

  const LISTOFTYPES = ["title", "name", "text"];

  await pg.schema.hasTable('type').then(async (exists) => {
    if (!exists) {
      await pg.schema
        .createTable('type', (table) => {
          table.increments();
          table.uuid('uuid');
          table.string('type');
          table.integer('typeID');
        }).then(async () => {
          console.log('created types');
          for (let i = 0; i < LISTOFTYPES.length; i++) {
            const uuid = Helpers.generateUUID();
            await pg.table('type').insert({
              uuid,
              type: LISTOFTYPES[i],
              typeID: i
            })
          }
        });


    }
  });
}
initialiseTables()


module.exports = app;